packages/flutter_tools/lib/src/asset.dart

```diff
@@ -0,0 +1,420 @@
+        manifestDescriptor['uses-material-design'];
+
+    for (_Asset asset in assetVariants.keys) {
+      AssetBundleEntry assetEntry = _createAssetEntry(asset);
+      if (assetEntry == null)
+        return 1;
+      entries.add(assetEntry);
+
+      for (_Asset variant in assetVariants[asset]) {
+        AssetBundleEntry variantEntry = _createAssetEntry(variant);
+        if (variantEntry == null)
+          return 1;
+        entries.add(variantEntry);
+      }
+    }
+
+    List<_Asset> materialAssets = <_Asset>[];
+    if (usesMaterialDesign) {
+      materialAssets.addAll(_getMaterialAssets(_kFontSetMaterial));
+      if (includeRobotoFonts)
+        materialAssets.addAll(_getMaterialAssets(_kFontSetRoboto));
+    }
+    for (_Asset asset in materialAssets) {
+      AssetBundleEntry assetEntry = _createAssetEntry(asset);
+      if (assetEntry == null)
+        return 1;
+      entries.add(assetEntry);
+    }
+
+    entries.add(_createAssetManifest(assetVariants));
+
+    AssetBundleEntry fontManifest =
+        _createFontManifest(manifestDescriptor, usesMaterialDesign, includeRobotoFonts);
+    if (fontManifest != null)
+      entries.add(fontManifest);
+
+    // TODO(ianh): Only do the following line if we've changed packages
+    entries.add(await _obtainLicenses(packageMap, assetBasePath));
+
+    return 0;
+  }
+
+  void dump() {
+    print('Dumping AssetBundle:');
+    for (AssetBundleEntry entry in entries) {
+      print(entry.archivePath);
+    }
+  }
+}
+
+class _Asset {
+  _Asset({ this.base, String assetEntry, this.relativePath, this.source }) {
+    this._assetEntry = assetEntry;
+  }
+
+  String _assetEntry;
+
+  final String base;
+
+  /// The entry to list in the generated asset manifest.
+  String get assetEntry => _assetEntry ?? relativePath;
+
+  /// Where the resource is on disk relative to [base].
+  final String relativePath;
+
+  final String source;
+
+  File get assetFile {
+    return new File(source != null ? '$base/$source' : '$base/$relativePath');
+  }
+
+  bool get assetFileExists => assetFile.existsSync();
+
+  /// The delta between what the assetEntry is and the relativePath (e.g.,
+  /// packages/flutter_gallery).
+  String get symbolicPrefix {
+    if (_assetEntry == null || _assetEntry == relativePath)
+      return null;
+    int index = _assetEntry.indexOf(relativePath);
+    return index == -1 ? null : _assetEntry.substring(0, index);
+  }
+
+  @override
+  String toString() => 'asset: $assetEntry';
+}
+
+Map<String, dynamic> _readMaterialFontsManifest() {
+  String fontsPath = path.join(path.absolute(Cache.flutterRoot),
+      'packages', 'flutter_tools', 'schema', 'material_fonts.yaml');
+
+  return loadYaml(new File(fontsPath).readAsStringSync());
+}
+
+final Map<String, dynamic> _materialFontsManifest = _readMaterialFontsManifest();
+
+List<Map<String, dynamic>> _getMaterialFonts(String fontSet) {
+  return _materialFontsManifest[fontSet];
```

Comment:
Nit: another place I'd prefer `=>` but like I said I'm kind of weirdly enamored with the form.